03-path-application-development/310-chaos-engineering/readme.adoc

```diff
@@ -0,0 +1,281 @@
+
+In this experiment's case there is _also_ a `probe` in the method. Probes in an experiment's method give us a chance to collate more information as the real-world variables are being manipulated by the experiment. The `probe` here extends the output of the experiment with the logs from pods labelled with `app==webapp-pod`.
+
+=== Rollbacks
+
+It is sometimes useful to supply an additional set of actions at the end of an experiment so that any actions in the method that were undertaken can be explicitly reversed. These are contained in a `rollback` section, but as Kubernetes will recover from this experiment's actions anyway there are no rollback actions required in this case:
+
+[source, JSON]
+----
+    "rollbacks": []
+}
+----
+
+This completes the experiment definition.
+
+== Running the Experiment
+
+With your cluster running you will first need to ensure you populate the `WEBAPP_URL` environment variable with the URL of your cluster's `webapp-service` endpoint.
+
+  $ export WEBAPP_URL="http://$(kubectl get svc/webapp-service -o jsonpath={.status.loadBalancer.ingress[0].ip})/"
+
+Now you can run the link:./experiments/experiment.json[experiment] using the link:http://chaostoolkit.org/[Chaos Toolkit's] `chaos run` command:
+
+  $ chaos run experiment.json
+  [2018-03-10 14:42:38 INFO] Validating the experiment's syntax
+  [2018-03-10 14:42:38 INFO] Experiment looks valid
+  [2018-03-10 14:42:38 INFO] Running experiment: Terminate the greeting service should not impact users
+  [2018-03-10 14:42:38 INFO] Steady state hypothesis: Services are all available and healthy
+  [2018-03-10 14:42:38 INFO] Probe: application-should-be-alive-and-healthy
+  [2018-03-10 14:42:38 INFO] Probe: application-must-respond-normally
+  [2018-03-10 14:42:39 INFO] Steady state hypothesis is met!
+  [2018-03-10 14:42:39 INFO] Action: terminate-greeting-service
+  [2018-03-10 14:42:40 INFO] Probe: fetch-application-logs
+  [2018-03-10 14:42:41 INFO] Steady state hypothesis: Services are all available and healthy
+  [2018-03-10 14:42:41 INFO] Probe: application-should-be-alive-and-healthy
+  [2018-03-10 14:42:42 INFO] Probe: application-must-respond-normally
+  [2018-03-10 14:42:45 ERROR]   => failed: activity took too long to complete
+  [2018-03-10 14:42:45 CRITICAL] Steady state probe 'application-must-respond-normally' is not in the given tolerance so failing this experiment
+  [2018-03-10 14:42:45 INFO] Let's rollback...
+  [2018-03-10 14:42:45 INFO] No declared rollbacks, let's move on.
+  [2018-03-10 14:42:45 INFO] Experiment ended with status: failed
+
+The output of the `chaos run` command shows that the experiment was run _but_ there is a weakness in the system. When the `greeting-service` is killed the `webapp-service` endpoint returns a response that is greater than the 3 seconds allowed as the tolerance for the system to be observed as still in steady-state.
+
```

Comment:
A guidance on how to fix this weakness and then run the test again with a passed status would be useful